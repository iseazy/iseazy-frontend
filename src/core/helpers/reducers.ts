import { Card } from "../../hooks/AppContext";


export const updateFlippedCardsArray = (array: Card[], value: Card) => {
    const existsAlready = array.some((item) => item.id === value.id)
    if(existsAlready) return array;
    return [...array,value];
}