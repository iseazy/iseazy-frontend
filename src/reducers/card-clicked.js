import { REPEATED_CARDS } from '../constants';
import { flipCardById, isCardClicked, isCardFlipped, isCardNotClicked } from '../utils';

export const cardClicked = (state, id) => { 
    const { cards } = state;
    const clickedCard = cards.find(card => card.id === id);

    let updatedCards = cards;

    if ([cards.filter(isCardClicked).length < REPEATED_CARDS,
        !isCardFlipped(clickedCard),
        isCardNotClicked(clickedCard)].every(Boolean)) {
        updatedCards = cards.map(flipCardById(id))
    }

    return {
        ...state,
        cards: updatedCards
    } 
};