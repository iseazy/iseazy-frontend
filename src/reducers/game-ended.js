import { GAME_STATE } from '../constants';

export const gameEnded = (state) => ({
    ...state,
    gameState: GAME_STATE.END,
    totalTime: (Date.now() - state.startTime)/1000
});
