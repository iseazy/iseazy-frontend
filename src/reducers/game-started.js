import { CARDS_SRC, GAME_STATE } from '../constants';
import { createDeck } from '../utils';

export const gameStarted = (state) => ({
    ...state,
    gameState: GAME_STATE.PLAY,
    cards: createDeck(CARDS_SRC),
    startTime: Date.now()
});
