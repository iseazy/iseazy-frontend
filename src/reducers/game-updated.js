import { REPEATED_CARDS } from '../constants';
import { areClickedCardsPaired, isCardClicked, unclickCard, unflipClickedCard } from '../utils';

export const gameUpdated = (state) => {
    const { cards } = state;
    const clickedCards = cards.filter(isCardClicked);

    let updatedCards = cards;

    if (clickedCards.length >= REPEATED_CARDS) {
        updatedCards = areClickedCardsPaired(clickedCards)
            ? cards.map(unclickCard)
            : cards.map(unflipClickedCard);
    }

    return {
        ...state,
        cards: updatedCards
    }
};
