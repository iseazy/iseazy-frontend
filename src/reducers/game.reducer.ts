import { GameAction, GameActionTypes } from '../actions/game.actions'
import { Game } from '../entities/game.entity'
import { initializeGame } from '../usecases/game'

export function gameReducer(state: Game, action: GameAction): Game {
  switch (action.type) {
    case GameActionTypes.START_GAME:
      return { ...state, startTime: Date.now() }
    case GameActionTypes.END_GAME:
      return { ...state, endTime: Date.now() }
    case GameActionTypes.RESTART_GAME:
      return initializeGame()
    case GameActionTypes.CLEAN_SELECTED_ITEMS:
      return { ...state, selectedItems: [] }
    case GameActionTypes.UPDATE_MATCHED_ITEMS:
      if (state.matchedItems.includes(action.payload)) return state
      return { ...state, matchedItems: [...state.matchedItems, action.payload] }
    case GameActionTypes.UPDATE_SELECTED_ITEMS:
      if (state.selectedItems.includes(action.payload)) return state
      return { ...state, selectedItems: [...state.selectedItems, action.payload] }
    default:
      return state
  }
}
