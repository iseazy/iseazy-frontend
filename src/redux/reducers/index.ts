import { combineReducers } from "redux";
import gameReducer from "./gameReducer";

export const rootReducer = combineReducers({
  gameReducer,
});

export type RootState = ReturnType<typeof rootReducer>