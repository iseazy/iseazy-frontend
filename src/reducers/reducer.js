import { GAME_ACTIONS } from '../constants';
import { cardClicked } from './card-clicked';
import { gameEnded } from './game-ended';
import { gameStarted } from './game-started';
import { gameUpdated } from './game-updated';

export const reducer = (state, action) => {
    switch (action.type) {
        case GAME_ACTIONS.START:
            return gameStarted(state);
        case GAME_ACTIONS.CLICK:
            return cardClicked(state, action.payload);
        case GAME_ACTIONS.UPDATE:
            return gameUpdated(state);
        case GAME_ACTIONS.END:
            return gameEnded(state);
        default:
            throw new Error();
    }
};