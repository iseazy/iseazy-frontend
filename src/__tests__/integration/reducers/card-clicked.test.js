import { cardClicked } from '../../../reducers/card-clicked';

describe('cardClicked()', () => {
    it('should not produce any effect on an already clicked card', () => {
        expect(cardClicked({ cards: mockDeck }, mockClickedId)).toStrictEqual({ cards: mockDeck });
    });

    it('should not produce any effect on an already paired card', () => {
        expect(cardClicked({ cards: mockDeck }, mockPairedId)).toStrictEqual({ cards: mockDeck });
    });

    it('should result flip unflipped card', () => {
        expect(cardClicked({ cards: mockDeck }, mockUnflippedId)).toMatchSnapshot();
    });
});
