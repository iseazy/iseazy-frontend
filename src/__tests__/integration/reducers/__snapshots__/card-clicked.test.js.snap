// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cardClicked() should result flip unflipped card 1`] = `
Object {
  "cards": Array [
    Object {
      "clicked": false,
      "flipped": true,
      "id": 1,
      "sharedId": 123,
      "src": "testSourceA",
    },
    Object {
      "clicked": false,
      "flipped": true,
      "id": 2,
      "sharedId": 123,
      "src": "testSourceA",
    },
    Object {
      "clicked": true,
      "flipped": true,
      "id": 3,
      "sharedId": 456,
      "src": "testSourceB",
    },
    Object {
      "clicked": true,
      "flipped": true,
      "id": 4,
      "sharedId": 456,
      "src": "testSourceB",
    },
    Object {
      "clicked": false,
      "flipped": false,
      "id": 5,
      "sharedId": 789,
      "src": "testSourceC",
    },
    Object {
      "clicked": false,
      "flipped": false,
      "id": 6,
      "sharedId": 789,
      "src": "testSourceC",
    },
  ],
}
`;
